File: src/server/model/order.go
package model

import (
	"fmt"
	"time"

	printfulmodel "github.com/baldurstod/go-printful-sdk/model"
	"github.com/shopspring/decimal"
)

type Order struct {
	ID                 string                       `json:"id" bson:"id"`
	Currency           string                       `json:"currency" bson:"currency"`
	ShippingAddress    Address                      `json:"shipping_address" bson:"shipping_address"`
	BillingAddress     Address                      `json:"billing_address" bson:"billing_address"`
	SameBillingAddress bool                         `json:"same_billing_address" bson:"same_billing_address"`
	Items              []OrderItem                  `json:"items" bson:"items"`
	ShippingInfos      []printfulmodel.ShippingRate `json:"shipping_infos" bson:"shipping_infos"`
	TaxInfo            TaxInfo                      `json:"tax_info" bson:"tax_info"`
	PercentDiscount    decimal.Decimal              `json:"percent_discount" bson:"percent_discount"`
	PriceDiscount      decimal.Decimal              `json:"price_discount" bson:"price_discount"`
	ShippingMethod     string                       `json:"shipping_method" bson:"shipping_method"`
	ItemsPrice         decimal.Decimal              `json:"items_price"`
	DiscountPrice      decimal.Decimal              `json:"discount_price"`
	ShippingPrice      decimal.Decimal              `json:"shipping_price"`
	TaxPrice           decimal.Decimal              `json:"tax_price"`
	TotalPrice         decimal.Decimal              `json:"total_price"`
	PrintfulOrderID    string                       `json:"printful_order_id" bson:"printful_order_id"`
	PaypalOrderID      string                       `json:"paypal_order_id" bson:"paypal_order_id"`
	Status             string                       `json:"status" bson:"status"`
	DateCreated        time.Time                    `json:"date_created" bson:"date_created"`
	DateUpdated        time.Time                    `json:"date_updated" bson:"date_updated"`
}

func NewOrder() Order {
	percent := decimal.NewFromFloat32(0.1)
	return Order{ShippingInfos: make([]printfulmodel.ShippingRate, 0), SameBillingAddress: true, PercentDiscount: percent, Items: make([]OrderItem, 0)}
}

func (order *Order) GetShippingInfo(shippingMethod string) *printfulmodel.ShippingRate {
	for _, shippingRate := range order.ShippingInfos {
		if shippingMethod == shippingRate.Shipping {
			return &shippingRate
		}
	}
	return nil
}

func (order *Order) GetItemsPrice() *decimal.Decimal {
	price := decimal.Decimal{}
	for _, item := range order.Items {
		price = price.Add(decimal.NewFromInt(int64(item.Quantity)).Mul(item.GetRetailPrice()))
	}

	price = price.Round(2)
	return &price
}

func (order *Order) GetShippingPrice() (*decimal.Decimal, error) {
	if order.ShippingMethod == "" {
		return &decimal.Decimal{}, nil
	}
	shippingInfo := order.GetShippingInfo(order.ShippingMethod)
	if shippingInfo == nil {
		return nil, fmt.Errorf("shipping info for method %s not found", order.ShippingMethod)
	}
	price, err := decimal.NewFromString(shippingInfo.Rate)
	if err != nil {
		return nil, fmt.Errorf("unable to convert shipping rate %s to decimal: %w", shippingInfo.Rate, err)
	}
	price = price.Round(2)
	return &price, nil
}

func (order *Order) GetTaxPrice() (*decimal.Decimal, error) {
	taxRate := decimal.NewFromFloat(order.TaxInfo.Rate)
	price := order.GetItemsPrice().Mul(taxRate)

	if order.TaxInfo.ShippingTaxable {
		shippingPrice, err := order.GetShippingPrice()
		if err != nil {
			return nil, err
		}
		price = price.Add(shippingPrice.Mul(taxRate))
	}

	price = price.Round(2)
	return &price, nil
}

func (order *Order) GetTotalPrice() (*decimal.Decimal, error) {
	percentOff, _ := decimal.NewFromString(order.PercentDiscount.String())
	priceDiscount, _ := decimal.NewFromString(order.PriceDiscount.String())

	shippingPrice, err := order.GetShippingPrice()
	if err != nil {
		return nil, err
	}

	taxPrice, err := order.GetTaxPrice()
	if err != nil {
		return nil, err
	}

	price := order.GetItemsPrice().Mul(decimal.NewFromInt(1).Sub(percentOff)).Sub(priceDiscount).Add(*shippingPrice).Add(*taxPrice)

	price = price.Round(2)
	return &price, nil
}

File: src/client/ts/model/order.ts
import { JSONArray, JSONObject } from 'harmony-types';
import { roundPrice } from '../common';
import { DEFAULT_SHIPPING_METHOD } from '../constants';
import { OrderJSON } from '../responses/order';
import { Address } from './address';
import { OrderItem } from './orderitem';
import { ShippingInfo } from './shippinginfo';
import { TaxInfo } from './taxinfo';

export class Order {
	#id = '';
	#currency = 'USD';
	#creationTime = 0;
	#shippingAddress = new Address();
	#billingAddress = new Address();
	#sameBillingAddress = true;
	#items: OrderItem[] = [];
	#shippingInfos = new Map<string, ShippingInfo>();
	#taxInfo = new TaxInfo();
	#shippingMethod = DEFAULT_SHIPPING_METHOD;
	#printfulOrder: any/*TODO: improve type*/;
	#paypalOrderId = '';
	#status = 'created';

	set items(items) {
		this.#items.length = 0;
		if (items) {
			for (const item of items) {
				this.#items.push(item);
			}
		}
	}

	get items(): OrderItem[] {
		return this.#items;
	}

	get id(): string {
		return this.#id;
	}

	get shippingAddress(): Address {
		return this.#shippingAddress;
	}

	get billingAddress(): Address {
		return this.#sameBillingAddress ? this.#shippingAddress : this.#billingAddress;
	}

	get sameBillingAddress(): boolean {
		return this.#sameBillingAddress;
	}

	set sameBillingAddress(sameBillingAddress) {
		this.#sameBillingAddress = sameBillingAddress;
	}

	getSameBillingAddress(): boolean {
		return this.#sameBillingAddress;
	}

	setSameBillingAddress(sameBillingAddress: boolean): void {
		this.#sameBillingAddress = sameBillingAddress;
	}

	addShippingInfo(shippingInfo: ShippingInfo): void {
		this.#shippingInfos.set(shippingInfo.shipping, shippingInfo);
	}

	get shippingInfos(): Map<string, ShippingInfo> {
		return this.#shippingInfos;
	}

	get shippingInfo(): ShippingInfo | null {
		return this.#shippingInfos.get(this.#shippingMethod) ?? this.#shippingInfos.get(DEFAULT_SHIPPING_METHOD) ?? null;
	}

	set taxInfo(taxInfo) {
		this.#taxInfo = taxInfo;
	}

	get taxInfo(): TaxInfo {
		return this.#taxInfo;
	}

	set currency(currency) {
		this.#currency = currency;
	}

	get currency(): string {
		return this.#currency;
	}

	set creationTime(creationTime) {
		this.#creationTime = creationTime;
	}

	get creationTime(): number {
		return this.#creationTime;
	}

	set paypalOrderId(paypalOrderId) {
		this.#paypalOrderId = paypalOrderId;
	}

	get paypalOrderId(): string {
		return this.#paypalOrderId;
	}

	set status(status) {
		this.#status = status;
	}

	get status(): string {
		return this.#status;
	}

	get itemsPrice(): number {
		let price = 0;
		for (const item of this.#items) {
			price += item.getSubtotal();
		}
		return this.roundPrice(price);
	}

	get shippingPrice(): number {
		if (this.shippingInfo) {
			return this.roundPrice(Number(this.shippingInfo.rate));
		}
		return 0;
	}

	get taxPrice(): number {
		if (this.shippingInfo && this.#taxInfo) {
			return this.roundPrice(this.itemsPrice * this.#taxInfo.rate + Number(this.shippingInfo.rate) * this.#taxInfo.rate * (this.#taxInfo.shippingTaxable ? 1 : 0));
		}
		return 0;
	}

	get totalPrice(): number | undefined {
		if (this.shippingInfo && this.#taxInfo) {
			return this.itemsPrice + this.shippingPrice + this.taxPrice;
		}
	}

	get shippingMethod(): string {
		return this.#shippingMethod;
	}

	set shippingMethod(shippingMethod) {
		this.#shippingMethod = shippingMethod;
	}

	get externalId() {
		return this.#printfulOrder?.external_id;
	}

	roundPrice(price: number): number {
		return roundPrice(this.#currency, price);
	}

	fromJSON(json: OrderJSON): void {
		this.#id = json.id;
		this.#currency = json.currency;
		this.#creationTime = json.date_created;
		this.#shippingAddress.fromJSON(json.shipping_address as JSONObject);
		this.#billingAddress.fromJSON(json.billing_address as JSONObject);
		this.#sameBillingAddress = json.same_billing_address;
		this.#items = [];
		if (json.items) {
			(json.items as JSONArray).forEach(
				(item) => {
					const orderItem = new OrderItem();
					orderItem.fromJSON(item as JSONObject);
					this.#items.push(orderItem);
				}
			);
		}

		this.#shippingInfos.clear();
		const shippingInfos = json.shipping_infos;
		if (shippingInfos) {
			for (const shippingInfoJSON of shippingInfos) {
				const shippingInfo = new ShippingInfo();
				shippingInfo.fromJSON(shippingInfoJSON as JSONObject);
				this.addShippingInfo(shippingInfo);
			}
		}

		this.#taxInfo.fromJSON(json.tax_info as JSONObject);
		this.#shippingMethod = json.shipping_method;
		this.#printfulOrder = json.printful_order_id;
		this.#paypalOrderId = json.paypal_order_id;
		this.#status = json.status;
	}

	toJSON() {
		const itemsJSON: JSONArray = [];
		this.#items.forEach(item => itemsJSON.push(item.toJSON()));

		return {
			id: this.id,
			currency: this.currency,
			creationTime: this.creationTime,
			shipping_address: this.shippingAddress.toJSON(),
			billing_address: this.billingAddress.toJSON(),
			same_billing_address: this.sameBillingAddress,
			items: itemsJSON,
			shipping_infos: this.#shippingInfos,
			taxInfo: this.taxInfo.toJSON(),
			shippingMethod: this.shippingMethod,
			printfulOrder: this.#printfulOrder,
			paypalOrderId: this.paypalOrderId,
			status: this.status,
			priceBreakDown: {
				itemsPrice: this.itemsPrice,
				shippingPrice: this.shippingPrice,
				taxPrice: this.taxPrice,
				totalPrice: this.totalPrice,
			}
		}
	}
}
